option to kill jobs if one of its outputs is removed

jwbargsten opened the pull request:
usually, all output datasets need to be removed from the history before a job is killed by Galaxy. With the new option "delete_output_dataset_kills_job" activated it is sufficient to just delete one output dataset to kill the job and deleted all other associated output datasets of the job.

Rationale:
The users in our lab have no straight forward possibility to kill a job, especially with hidden output datasets. Usually the steps to kill a job are to delete all (visible) outputs, make hidden outputs visible and also delete them. This "concept" is easily misunderstood, causing many "zombie" jobs on our (cluster) system that block resources.

jwbargsten committed:
option to kill jobs if one of its outputs is removed

usually, all output datasets need to be removed from the history
before a job is killed by Galaxy. With the new option
"delete_output_dataset_kills_job" activated it is sufficient to
just delete one output dataset to kill the job and deleted all
other associated output datasets of the job.

mvdbeek commented:
I like this in principle, but I'm a little worried about confusing users that use multiple galaxy servers and that may be used to one or the other behavior. What would be great is if we could allow users to interactively choose whether they want to cancel the job when they delete a dataset that is in running state. I am not opposed to merging this in the meantime though.

mvdbeek commented:
@galaxybot test this

mvdbeek commented on lib/galaxy/managers/datasets.py:
```diff
@@ -341,7 +341,7 @@ def stop_creating_job(self, dataset_assoc):
             if not job.finished:
                 # Are *all* of the job's other output datasets deleted?
-                if job.check_if_output_datasets_deleted():
+                if job.check_if_output_datasets_deleted() or self.app.config.delete_output_dataset_kills_job:
```
Can you adjust the comment in line 343 and the docstring of this function to reflect the new behavior ?

nsoranzo commented on config/galaxy.ini.sample:
```diff
@@ -1256,6 +1256,11 @@ use_interactive = True
+# Usually, a job is only killed if all output datasets are deleted in the history.
+# This option let's galaxy kill the job after only one of the jobs output datasets
+# is deleted
+#delete_output_dataset_kills_job = True
```
The value in the commented option should be the default, which below is correctly defined as `False`.

jwbargsten committed:
Improved documentation and comment in function

..."stop_creating_job" to reflect changes introduced by new
delete_output_dataset_kills_job configuration option

nsoranzo commented:
> I like this in principle, but I'm a little worried about confusing users that use multiple galaxy servers and that may be used to one or the other behavior. What would be great is if we could allow users to interactively choose whether they want to cancel the job when they delete a dataset that is in running state.

I also think that this would be better implemented on the UI side, having a dataset deleted via the API delete the whole job doesn't feel right at all.

jwbargsten committed:
fix: config option default not correct in sample config

natefoo commented:
I did a tiny amount of work on a "job-oriented view" of the history a while ago for this sort of purpose. I didn't get very far because I don't understand the UI and got pulled away before I could really get to learning it, but if anyone's interested in looking at it, it's here: https://github.com/natefoo/galaxy/tree/job-view

mvdbeek commented:
What would be even cooler is if a job belongs to a workflow, that the user can cancel the workflow.

bgruening commented:
Our users remove frequently outputs they don't need (logs, unneeded outputs etc) and they don't expect jobs to be cancelled.
My guess is that the underlying issue here is the hidden datasets, that are not reachable.
@jwbargsten in which cases are your datasets hidden? Is this with collections?

martenson commented:
@galaxybot test this

jwbargsten commented:
@bgruening Probably I'm outing myself as a person that is using Galaxy not as intended, but we have jobs that are run on our cluster via a small wrapper script in Galaxy. I'll not go into details here, but the main point is that the raw log info and config from the cluster is made available as hidden output of the tools. This gives me (as admin) much better angle at debugging errors.

I probably could dump this info into stderr and tell the users that they need to remove all outputs, but I thought I give it a try by submitting a pull request. Even if my request is not incorporated, I see that this PR spawned a nice discussion. I also think that in the long run a UI-based solution might be the way to go, but this requires much more sophistication.

bgruening commented:
@jwbargsten I don't want to discourage you here. One of the strength of Galaxy is that it is nicely hackable :)

I'm just worried to expose these to all users.

>I'll not go into details here

Maybe you should, or on gitter, so we can find a nice way around this hack?

For example, if you already have such a wrapper to submit your jobs, have a look at [sentry](https://github.com/getsentry/sentry). You could send these information directly to a sentry server without having your users to deal.
But maybe we can write a job_runner for you cluster and you don't need this hack at all?

mvdbeek commented:
> For example, if you already have such a wrapper to submit your jobs, have a look at sentry. You could send these information directly to a sentry server without having your users to deal.
> But maybe we can write a job_runner for you cluster and you don't need this hack at all?
> Merge state

We just need to fix this in the UI (this is very useful functionality), creating job runners or needing sentry seems like a worse hack and then we might as well merge this PR, which is almost right IMO.

bgruening commented:
@mvdbeek I was referring to: `I'll not go into details here, but the main point is that the raw log info and config from the cluster is made available as hidden output of the tools.`

Let me know if Sentry or a dedicated job_runner is a hack in comparison to this.

mvdbeek commented:
Ah, I didn't get that.

martenson commented:
We have pushed this to 18.05 since there seems to be ideas about a client-side solution for this.

jmchilton commented:
Pushing again for same reason, marking as WIP because of conflicts.
